--- src/Plugin/Validation/Constraint/FieldValidationConstraintValidator.php
+++ src/Plugin/Validation/Constraint/FieldValidationConstraintValidator.php
@@ -94,6 +94,7 @@

     // Property level validation
     if ($items->count() !== 0) {
+      $property_path = $this->context->getPropertyPath();
       foreach ($items as $delta => $item) {
         // You can hard code configuration or you load from settings.
         foreach ($rules_property as $rule) {
@@ -112,6 +113,7 @@
               $real_constraint = $constraint_manager->createInstance($constraint_name, $constraint_options);
               $constraint_validator_factory =  new ConstraintValidatorFactory($class_resolver);
               $validator = $constraint_validator_factory->getInstance($real_constraint);
+              $this->context->setNode($this->context->getValue(), $this->context->getObject(), $this->context->getMetadata(), $property_path . '.' . $delta . '.' . $column);
               $validator->initialize($this->context);
               $validator->validate($value, $real_constraint);
             }
