diff --git a/src/BookManager.php b/src/BookManager.php
--- a/src/BookManager.php
+++ b/src/BookManager.php
@@ -763,24 +763,15 @@
       // Allow book-specific theme overrides.
       $element['attributes'] = new Attribute();
       $element['title'] = $data['link']['title'];
-      $route = 'entity.node.canonical';
-      $route_parameters = [
-        'node' => $data['link']['nid'],
-      ];
       $route_options = [
         'language' => $this->languageManager->getCurrentLanguage(LanguageInterface::TYPE_CONTENT),
       ];
-      $element['url'] = Url::fromRoute($route, $route_parameters, $route_options);
+      $element['url'] = $data['link']['entity']->toUrl('canonical', $route_options);
       $node = $this->route_match->getParameter('node');
       // Check if item belongs to the current page.
       if ($node instanceof NodeInterface && $node->id() === $data['link']['nid']) {
-        $link_options = [
-          'attributes' => [
-            'class' => [
-              'is-active',
-            ],
-          ],
-        ];
+        $link_options = $element['url']->getOptions();
+        $link_options['attributes']['class'][] = 'is-active';
         $element['url']->setOptions($link_options);
       }

@@ -1209,6 +1201,7 @@
       // The node label will be the value for the current language.
       $node = $this->entityRepository->getTranslationFromContext($node);
       $link['title'] = $node->label();
+      $link['entity'] = $node;
       $link['options'] = [];
     }
     return $link;

diff --git a/src/Hook/BookThemeHooks.php b/src/Hook/BookThemeHooks.php
--- a/src/Hook/BookThemeHooks.php
+++ b/src/Hook/BookThemeHooks.php
@@ -116,7 +116,7 @@
     // Provide extra variables for themers. Not needed by default.
     $variables['book_id'] = $book_link['bid'];
     $variables['book_title'] = $book_top['title'] ?? '';
-    $variables['book_url'] = Url::fromRoute('entity.node.canonical', ['node' => $book_link['bid']])->toString();
+    $variables['book_url'] = $book_top['entity']->toUrl('canonical')->toString();
     $variables['current_depth'] = $book_link['depth'] ?? 0;
     $variables['tree'] = '';

@@ -128,7 +128,7 @@
       $build = [];

       if ($prev = $book_outline->prevLink($book_link)) {
-        $prev_href = Url::fromRoute('entity.node.canonical', ['node' => $prev['nid']])->toString();
+        $prev_href = $prev['entity']->toUrl('canonical')->toString();
         $build['#attached']['html_head_link'][][] = [
           'rel' => 'prev',
           'href' => $prev_href,
@@ -138,7 +138,7 @@
       }

       if ($book_link['pid'] && $parent = $book_manager->loadBookLink($book_link['pid'])) {
-        $parent_href = Url::fromRoute('entity.node.canonical', ['node' => $book_link['pid']])->toString();
+        $parent_href = $book_top['entity']->toUrl('canonical')->toString();
         $build['#attached']['html_head_link'][][] = [
           'rel' => 'up',
           'href' => $parent_href,
@@ -148,7 +148,7 @@
       }

       if ($next = $book_outline->nextLink($book_link)) {
-        $next_href = Url::fromRoute('entity.node.canonical', ['node' => $next['nid']])->toString();
+        $next_href = $next['entity']->toUrl('canonical')->toString();
         $build['#attached']['html_head_link'][][] = [
           'rel' => 'next',
           'href' => $next_href,
