--- a/core/modules/taxonomy/src/Form/OverviewTerms.php
+++ b/core/modules/taxonomy/src/Form/OverviewTerms.php
@@ -277,6 +277,7 @@
       '#empty' => $empty,
       '#header' => [
         'term' => $this->t('Name'),
+        'description' => $this->t('Description'),
         'status' => $this->t('Status'),
         'operations' => $this->t('Operations'),
         'weight' => !$operations_access ? $this->t('Weight') : NULL,
@@ -290,6 +291,7 @@
     foreach ($current_page as $key => $term) {
       $form['terms'][$key] = [
         'term' => [],
+        'description' => [],
         'status' => [],
         'operations' => [],
         'weight' => $update_tree_access->isAllowed() ? [] : NULL,
@@ -310,6 +312,7 @@
         '#title' => $term->getName(),
         '#url' => $term->toUrl(),
       ];
+      $form['terms'][$key]['description'] = ['#plain_text' => $term->getDescription()];
       $form['terms'][$key]['status'] = [
         '#type' => 'item',
         '#markup' => ($term->isPublished()) ? $this->t('Published') : $this->t('Unpublished'),
