--- a/src/Plugin/views/argument/GroupId.php
+++ b/src/Plugin/views/argument/GroupId.php
@@ -3,6 +3,7 @@
 namespace Drupal\group\Plugin\views\argument;

 use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Entity\EntityRepositoryInterface;
 use Drupal\views\Attribute\ViewsArgument;
 use Drupal\views\Plugin\views\argument\NumericArgument;
 use Symfony\Component\DependencyInjection\ContainerInterface;
@@ -18,6 +19,7 @@
     $plugin_id,
     $plugin_definition,
     protected EntityTypeManagerInterface $entityTypeManager,
+    protected EntityRepositoryInterface $entityRepository,
   ) {
     parent::__construct($configuration, $plugin_id, $plugin_definition);
   }
@@ -30,7 +32,8 @@
       $configuration,
       $plugin_id,
       $plugin_definition,
-      $container->get('entity_type.manager'),
+      $container->get('entity_type.manager'),
+      $container->get('entity.repository'),
     );
   }

@@ -38,11 +41,11 @@
    * Override the behavior of title(). Get the title of the group.
    */
-  public function titleQuery() {
+  public function titleQuery(): array {
     $titles = [];

     $groups = $this->entityTypeManager->getStorage('group')->loadMultiple($this->value);
     foreach ($groups as $group) {
-      $titles[] = $group->label();
+      $titles[] = $this->entityRepository->getTranslationFromContext($group)->label();
     }

     return $titles;
