--- a/src/Form/GroupPermissionsTypeSpecificForm.php
+++ b/src/Form/GroupPermissionsTypeSpecificForm.php
@@ -3,6 +3,7 @@
 namespace Drupal\group\Form;

 use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Extension\ModuleExtensionList;
 use Drupal\Core\Extension\ModuleHandlerInterface;
 use Drupal\Core\Form\FormStateInterface;
 use Drupal\group\Access\GroupPermissionHandlerInterface;
@@ -24,7 +25,7 @@
   public function __construct(
     protected EntityTypeManagerInterface $entityTypeManager,
     GroupPermissionHandlerInterface $permission_handler,
-    ModuleHandlerInterface $module_handler,
+    ModuleExtensionList $module_handler,
   ) {
     parent::__construct($permission_handler, $module_handler);
   }
@@ -36,7 +37,7 @@
     return new static(
       $container->get('entity_type.manager'),
       $container->get('group.permissions'),
-      $container->get('module_handler')
+      $container->get('extension.list.module')
     );
   }

